libstuff/SDeburr.cpp

```diff
@@ -0,0 +1,137 @@
+#include "SDeburr.h"
+
+#include <bit>
+#include <cstring>
+#include <string>
+
+#include <libstuff/libstuff.h>
+#include <libstuff/sqlite3.h>
+
+using namespace std;
+
+const char* SDeburr::unicodeToAscii(uint32_t codepoint) {
+    if (codepoint >= 0x00C0 && codepoint <= 0x017F) {
+        return SDeburr::UNICODE_TO_ASCII_MAP[codepoint - 0x00C0];
+    }
+
+    if (codepoint == 0x1E9E) {
+        return "SS"; // ẞ
+    }
+
+    return nullptr;
+}
+
+string SDeburr::deburr(const string& input) {
+    const unsigned char* input_bytes = reinterpret_cast<const unsigned char*>(input.c_str());
+    const size_t len = input.size();
+    string result;
+    result.reserve(len);
+    size_t i = 0;
+    while (i < len) {
+        // Count the leading ones in the current byte
+        // 0 leading ones -> ASCII character
+        // 1 leading one -> continuation byte
+        // 2-4 leading ones -> 2, 3, or 4-byte sequence
+        // 5+ leading ones -> invalid UTF-8
+        const int numLeadingOnes = countl_one(input_bytes[i]);
+
+        // Most common case: regular ASCII text
+        if (numLeadingOnes == 0) {
+            result.append(reinterpret_cast<const char*>(input_bytes + i), 1);
+            i++;
+            continue;
+        }
+
+        // Check for invalid UTF-8. If invalid, skip the current byte and continue.
+        // Cases we check for:
+        // - A continuation byte (1 leading one). This should be handled only as part of a multi-byte sequence.
+        // - More than 4 leading ones
+        // - Not enough bytes to form a valid sequence
+        if (numLeadingOnes == 1 || numLeadingOnes > 4 || (i + numLeadingOnes) > len) {
```

Comment:
Should this warn?

Reply:
I found something quite curious that I can't explain here. I added an `SWARN` in this block, and after doing that found that the benchmarks degraded by about 25%, even though the log is never printed to `/var/log/syslog`. I double-and-triple checked this - it's very reliable to reproduce. I also tried adding a regular `cout` in there and found that it wasn't printing anything.

So yeah, that's weird.

Reply:
The fact that logging macros can impact performance, even if they're not executed, suggests to me that they'd be better off as normal functions or perhaps inline function (though my reading suggests it's better to just let the compiler decide when functions should be inlined or not).